flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java

```diff
@@ -86,39 +98,38 @@ public void open(SerializationSchema.InitializationContext context) throws Excep
 
 	@Override
 	public ProducerRecord<byte[], byte[]> serialize(RowData consumedRow, @Nullable Long timestamp) {
-		final RowData physicalRow;
-		// shortcut if no metadata is required
-		if (!hasMetadata) {
-			physicalRow = consumedRow;
-		} else {
-			final int physicalArity = physicalFieldGetters.length;
-			final GenericRowData genericRowData = new GenericRowData(
-					consumedRow.getRowKind(),
-					physicalArity);
-			for (int i = 0; i < physicalArity; i++) {
-				genericRowData.setField(i, physicalFieldGetters[i].getFieldOrNull(consumedRow));
-			}
-			physicalRow = genericRowData;
+		// shortcut in case no input projection is required
+		if (keySerialization == null && !hasMetadata) {
+			final byte[] valueSerialized = valueSerialization.serialize(consumedRow);
+			return new ProducerRecord<>(
+					topic,
+					extractPartition(consumedRow, null, valueSerialized),
+					null,
+					valueSerialized);
 		}
 
-		final byte[] valueSerialized = valueSerialization.serialize(physicalRow);
-
-		final Integer partition;
-		if (partitioner != null) {
-			partition = partitioner.partition(physicalRow, null, valueSerialized, topic, partitions);
+		final byte[] keySerialized;
+		final RowKind kind = consumedRow.getRowKind();
+		if (keySerialization == null) {
+			keySerialized = null;
 		} else {
-			partition = null;
+			final RowData keyRow = createProjectedRow(consumedRow, kind, keyFieldGetters);
+			keySerialized = keySerialization.serialize(keyRow);
 		}
 
-		// shortcut if no metadata is required
-		if (!hasMetadata) {
-			return new ProducerRecord<>(topic, partition, null, null, valueSerialized);
+		final byte[] valueSerialized;
+		if (kind == RowKind.DELETE) {
+			valueSerialized = null;
+		} else {
+			final RowData valueRow = createProjectedRow(consumedRow, kind, valueFieldGetters);
+			valueSerialized = valueSerialization.serialize(valueRow);
```

Comment:
Should under `upsertMode == true`.

Reply:
If uspert mode is true, we should transfrom the delete/update-before message to tombstone message. But I am not sure what behavirour of the `DynamicKafkaSerializationSchema`? It has no means to discard the row it receives. If we have the same behaviour, I think it's useless to introduce a flag.